Google.Api.Gax.Grpc/Rest/PartialDecodingStreamReader.cs

```diff
@@ -158,72 +137,82 @@ public async Task<bool> MoveNext(CancellationToken cancellationToken)
         }
     }
 
+    private async Task<bool> MoveNextImpl(CancellationToken cancellationToken)
+    {
+        while (true)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            if (_queuedResponses.Count > 0)
+            {
+                Current = _queuedResponses.Dequeue();
+                return true;
+            }
+
+            if (_endOfData)
+            {
+                return false;
+            }
+
+            // We don't currently know what to do, so read more data and see whether that
+            // satisfies one of the above conditions. (We may need to loop multiple times.)
+            await ReadAndProcessData(cancellationToken).ConfigureAwait(false);
+        }
+    }
+
     /// <summary>
     /// Called when we don't have a queued response and haven't reached a terminal state (error or completed).
     /// </summary>
     /// <returns></returns>
     private async Task ReadAndProcessData(CancellationToken cancellationToken)
     {
-        try
-        {
-            var textReader = await AwaitWithCancellation(_textReaderTask, cancellationToken).ConfigureAwait(false);
+        var textReader = await AwaitWithCancellation(_textReaderTask, cancellationToken).ConfigureAwait(false);
 
-            var readTask = textReader.ReadAsync(_readBuffer, 0, _readBuffer.Length);
-            int charsRead = await AwaitWithCancellation(readTask, cancellationToken).ConfigureAwait(false);
+        var readTask = textReader.ReadAsync(_readBuffer, 0, _readBuffer.Length);
+        int charsRead = await AwaitWithCancellation(readTask, cancellationToken).ConfigureAwait(false);
 
-            if (charsRead == 0)
+        if (charsRead == 0)
```

Comment:
Just noticing this now, but I think this can be `charsRead < _readBuffer.Length`. We'd have to process the buffer before checking for this then though.
But it does save one extra read attempt.

Reply:
Nope - it's entirely valid for a reader to return fewer characters than the max, and still have more later. (Indeed, our tests would do this.)